[TexMap] Incorrect damage information produced for animations with blur

Scony opened the pull request:
#### ab86837116be8c4fced269c90d9cea2ff3e3204e
<pre>
[TexMap] Incorrect damage information produced for animations with blur
<a href="https://bugs.webkit.org/show_bug.cgi?id=277484">https://bugs.webkit.org/show_bug.cgi?id=277484</a>

Reviewed by Nikolas Zimmermann.

This change:
- adds a mechanism to collect damage from replica/filter/mask&apos;s
  overlap regions
- makes the damage being propagated always when dirty region changes
- removes two workarounds invalidating whole layer in some cases
so that damage from ongoing animations is colleced correctly now.

* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamageRecursive):
(WebCore::TextureMapperLayer::collectDamageSelfAndChildren):
(WebCore::TextureMapperLayer::collectDamageSelfChildrenReplicaFilterAndMask):
(WebCore::TextureMapperLayer::collectDamageSelfChildrenFilterAndMask):
(WebCore::TextureMapperLayer::computeConsolidatedOverlapRegionRects):
(WebCore::TextureMapperLayer::paintSelfChildrenFilterAndMask):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::setDamage):
(WebCore::CoordinatedPlatformLayer::flushCompositingState):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):

Canonical link: <a href="https://commits.webkit.org/289158@main">https://commits.webkit.org/289158@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46d51329d8642320937966fa1f01741651447988

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85510 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5216 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39917 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90610 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36524 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87570 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5397 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13196 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66458 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24272 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88530 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4136 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77642 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46740 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4012 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31904 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35593 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74699 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32754 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92225 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12836 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9410 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75146 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13057 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73475 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74277 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18352 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18542 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16977 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4929 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12848 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18242 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12655 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16094 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14429 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4db7327fd847b7a725b34a109959edaff5e90eb5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37707

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a2fbbe016dd9152c73c6698275036f8d7f120790. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37707

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bf02e29c57f950ef19e0fa3dc399dddbdd68baf4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37707

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9317ce52b4b4584972b7713b6c4cf3b90301a487. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37707

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0d25a88623d31e08e8bcabdf3019a47b7fc7ec2e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37707

Scony commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -432,6 +439,76 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
+            return true;
+        return false;
+    }();
+    if (hasFilterOrMask) {
+        if (m_state.replicaLayer) {
+            SetForScope scopedReplicaLayer(options.replicaLayer, this);
+            SetForScope scopedTransform(options.transform, options.transform);
+            options.transform.multiply(replicaTransform());
+            collectDamageSelfChildrenFilterAndMask(options, damage);
+        }
+        collectDamageSelfChildrenFilterAndMask(options, damage);
+    } else {
+        if (m_state.replicaLayer) {
+            SetForScope scopedReplicaLayer(options.replicaLayer, this);
+            SetForScope scopedTransform(options.transform, options.transform);
+            options.transform.multiply(replicaTransform());
+            collectDamageSelfAndChildren(options, damage);
+        }
+        collectDamageSelfAndChildren(options, damage);
+    }
```
deduplicate

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/83352f988d3034637866420039a8c15bd38c7c52. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/37707

aperezdc reviewed (commented):
Changes look reasonable to me but I think it would be good to have another pair of eyes take a look (in particular at the calculations), just in case. I left a small comment but it is nothing that would block landing this patch.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
```diff
@@ -539,10 +539,8 @@ void CoordinatedPlatformLayer::setDirtyRegion(Vector<IntRect, 1>&& dirtyRegion)
-    m_damage = WTFMove(damage);
+    if (m_damage != damage)
+        m_damage = WTFMove(damage);
     m_pendingChanges.add(Change::Damage);
```
You always want to add `Change::Damage` to `m_pendingChanges`, even if `damage` wasn't changed?

Scony replied:
Correct. The current implementation of damage handling in `TextureMapperLayer` relies on the fact that damage is set every frame, even if it haven't changed between frames. And this code here is being triggered by `GraphicsLayerCoordinated::updateDirtyRegion()` that is called only if `DirtyRegion` changed. So atm. that's expected.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -284,6 +287,7 @@ class TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLayer> {
 #if ENABLE(DAMAGE_TRACKING)
     bool m_damagePropagation { false };
     Damage m_damage;
+    FloatRect m_accumulatedOverlapRegionDamage { };
```
s/ { } /

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -432,6 +439,73 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
+    ASSERT(nonOverlapRegion.isEmpty());
+
+    auto rects = overlapRegion.rects();
+    static const size_t OverlapRegionConsolidationThreshold = 4;
```
Any good reason for the 4? I'd name it 'overlapRegion...'

nikolaszimmermann replied:
`constexpr size_t`?

Scony replied:
This code is a copy-paste from another place. I've extracted the common part now.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -432,6 +439,73 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
+        nonOverlapRegion
+    };
+    data.clipBounds.move(-options.offset);
+    computeOverlapRegions(data, options.transform, false);
```
Side-note: I dislike the third parameter -- unclear what it is.

nikolaszimmermann reviewed (approved):
I think the commit message should be improved... it fixes a whole class of damage computation issues that you summarized as 'removes two workarounds invalidating whole layer in some cases' :-) Animations aren't mentioned, where this is a big win.

Looks good!

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/46d51329d8642320937966fa1f01741651447988)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85510 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5216 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39917 "Build is in progress. Recent messages:Running configuration") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36524 "Build is in progress. Recent messages:Running configuration") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87570 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5397 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13196 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/90610 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/36524 "Build is in progress. Recent messages:Running configuration") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88530 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/5397 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/39917 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/90610 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/5397 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/39917 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35593 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/5397 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/39917 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92225 "Build is in progress. Recent messages:Printed configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12836 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/13196 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/92225 "Build is in progress. Recent messages:Printed configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13057 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/39917 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/92225 "Build is in progress. Recent messages:Printed configuration") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/55/builds/39917 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4929 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12848 "Build is in progress. Recent messages:OS: Sequoia (15.0.1), Xcode: 16.0") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18242 "Build is in progress. Recent messages:") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12655 "Build is in progress. Recent messages:OS: Sequoia (15.0.1), Xcode: 16.0") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16094 "Build is in progress. Recent messages:") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14429 "Build is in progress. Recent messages:OS: Sequoia (15.0.1), Xcode: 16.0") | | | </details>

Scony committed:
[TexMap] Incorrect damage information produced for animations with blur
https://bugs.webkit.org/show_bug.cgi?id=277484

Reviewed by Nikolas Zimmermann.

This change:
- adds a mechanism to collect damage from replica/filter/mask's
  overlap regions
- makes the damage being propagated always when dirty region changes
- removes two workarounds invalidating whole layer in some cases
so that damage from ongoing animations is colleced correctly now.

* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamageRecursive):
(WebCore::TextureMapperLayer::collectDamageSelfAndChildren):
(WebCore::TextureMapperLayer::collectDamageSelfChildrenReplicaFilterAndMask):
(WebCore::TextureMapperLayer::collectDamageSelfChildrenFilterAndMask):
(WebCore::TextureMapperLayer::computeConsolidatedOverlapRegionRects):
(WebCore::TextureMapperLayer::paintSelfChildrenFilterAndMask):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::setDamage):
(WebCore::CoordinatedPlatformLayer::flushCompositingState):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):

Canonical link: https://commits.webkit.org/289158@main

webkit-commit-queue commented:
Committed 289158@main (ab86837116be): <https://commits.webkit.org/289158@main>

Reviewed commits have been landed. Closing PR #37707 and removing active labels.
